python/src/iceberg/io/pyarrow.py

```diff
@@ -0,0 +1,154 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""FileIO implementation for reading and writing table files that uses pyarrow.fs
+
+This file contains a FileIO implementation that relies on the filesystem interface provided
+by PyArrow. It relies on PyArrow's `from_uri` method that infers the correct filesystem
+type to use. Theoretically, this allows the supported storage types to grow naturally
+with the pyarrow library.
+"""
+
+import os
+from typing import Union
+from urllib.parse import urlparse
+
+from pyarrow.fs import FileSystem, FileType
+
+from iceberg.io.base import FileIO, InputFile, InputStream, OutputFile, OutputStream
+
+
+class PyArrowFile(InputFile, OutputFile):
+    """A combined InputFile and OutputFile implementation that uses a pyarrow filesystem to generate pyarrow.lib.NativeFile instances
+
+    Args:
+        location(str): A URI or a path to a local file
+
+    Attributes:
+        location(str): The URI or path to a local file for a PyArrowFile instance
+        exists(bool): Whether the file exists or not
+        filesystem(pyarrow.fs.FileSystem): An implementation of the FileSystem base class inferred from the location
+
+    Examples:
+        >>> from iceberg.io.pyarrow import PyArrowFile
+        >>> input_file = PyArrowFile("s3://foo/bar.txt")
+        >>> file_content = input_file.open().read()  # Read the contents of the PyArrowFile instance
+        >>> output_file = PyArrowFile("s3://baz/qux.txt")
+        >>> output_file.create().write(b'foobytes')  # Write bytes to the PyArrowFile instance
+    """
+
+    def __init__(self, location: str):
+        parsed_location = urlparse(location)  # Create a ParseResult from the uri
+        if not parsed_location.scheme:  # If no scheme, assume the path is to a local file
+            self._filesystem, self._path = FileSystem.from_uri(os.path.abspath(location))
+        else:
+            self._filesystem, self._path = FileSystem.from_uri(location)  # Infer the proper filesystem
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        file = self._filesystem.open_input_file(self._path)
+        return file.size()
+
+    @property
```

Comment:
Why is this a property instead of a method? Since this will commonly make an RPC call and isn't cached, it seems like something we'd want people to think of as a method call.

Reply:
Good point, I switched this to a method here and also in the `InputFile` and `OutputFile` base classes.